### DISCUSSION TITLE
Is there a way to convert from mind mapping software to NW?

### DISCUSSION THREAD

**Discussion by xahodo** [vkbo/novelWriter#1450] [General]
I'm using [Freeplane](https://docs.freeplane.org/) (yes, it's their website, and it's open source) to create a mind map of my story, after figuring out that just jumping in doesn't work out for me.

It is a nice tool for mind mapping. It outputs an XML file which is stored on the local drive (no fancy cloud support). What would be really cool if I could take what I created in Freeplane and convert it to a novel structure, including all the right note folders, subfolders, tags, and synopsis readily provided from my mind map.

Is anybody willing to create a bash/python script in order to do what I want? Or is there a hidden feature in NW, kept away from us by a dark and evil conspiracy :D , to import such mind maps and convert them to NW projects?

**Comment by vkbo** [comment]
There is no hidden feature for import/export. So far, there is only one developer that provides a tool for converting projects between novelWriter and yWriter formats.

I'm working on a new system for building manuscripts from novelWriter projects, and I could possibly add an interface/API for importing them too. Since novelWriter's source code is very modular, and even more so in 2.0, you can in principle write a Python script that builds a project in memory from any data source and spit out a complete project.

I basically do this on the fly in the test suite. It isn't particularly complicated, but it is undocumented: https://github.com/vkbo/novelWriter/blob/4371d123ac3b3ea46225b9b36ddf5776f174ff8a/tests/tools.py#L157-L212

**Comment by peter88213** [comment]
On this occasion I notice that NovelWriter does not seem to handle any import from foreign formats.

I can think of a way using existing Python scripts, but it is a bit cumbersome and has some limitations. One important requirement: there are three levels (part, chapter, scene). Then you can export a novel outline as ODT from _Freemind_. Note texts on all levels are taken over as synopses. From the ODT document you can create a yWriter project as an intermediate file with [yW2OO](https://peter88213.github.io/yW2OO/), which can be transferred into a NovelWriter project with [yw2nw](https://peter88213.github.io/yw2nw/).  

Now the catch: 
- With _Freeplane_ version 1.9.7 it doesn't work because the exported ODT document is not structured with heading styles. 
- With _FreeMind_ version 1.0.1 it works fine, so you might want to use it at least for ODT export.  

![freemind_to_novelwriter](https://github.com/vkbo/novelWriter/assets/42072564/f5c1cd34-d37f-4927-afd9-c0d2e924e6a9)

If you want to run _yW2OO_ on Linux, please make sure the tk support for Python is installed. 

UPDATE:
As of version 5.3.0, _yW2OO_ can read ODT documents without meta.xml as created with FreeMind.

**Reply by peter88213** [reply]
Having not been satisfied with the ODT export of the mindmappers, I sat down today and programmed a parser for the *.mm* XML file format. 

The new [mm2yw7](https://peter88213.github.io/mm2yw7/index) file converter simplifies the conversion process from mindmap to NovelWriter project. Moreover, characters, locations and items are converted if they are arranged in the mind map in corresponding branches marked by specified icons. This works no matter whether the mindmap is created with FreeMind or Freeplane. 

Now the conversion process looks like this: 

![freemind_to_novelwriter](https://github.com/vkbo/novelWriter/assets/42072564/e96a2ebd-0d9c-4dda-9a67-b366441447fc)

An example with FreeMind:

![freemind01](https://github.com/vkbo/novelWriter/assets/42072564/d3b836a9-b170-43c8-8946-46207bf21493)

![mm2nw](https://github.com/vkbo/novelWriter/assets/42072564/b0e7e13c-9772-4962-8b1a-ec9a2336726a)

The *mm2yw7* Python script is still in alpha testing stage, and the documentation is not yet complete, but the results are already worth seeing.

**Reply by vkbo** [reply]
That looks really nice actually!

Let me know how it goes, and I can perhaps add a page on the website about third party tools. Currently, there's a mention on the [HowTo](https://novelwriter.io/docs/int_howto.html) page, but I want to move some of it out of the docs.

**Reply by peter88213** [reply]
Thanks. I guess it also would be feasible to combine [mm2yw7](https://peter88213.github.io/mm2yw7) and [yw2nw](https://peter88213.github.io/yw2nw/) to a single script, thus leaving out the yw7 file generation. However, this would mean updating yw2nw due to outdated libraries, merging the documentation, and maintaining one more GitHub repository. Also, I can't promise that I will continue to maintain the yw2nw code if the file format of novelWriter changes ever again. 

Anyway, it's all free and open source ;-)

**Reply by peter88213** [reply]
> That looks really nice actually!
> 
> Let me know how it goes, and I can perhaps add a page on the website about third party tools. Currently, there's a mention on the [HowTo](https://novelwriter.io/docs/int_howto.html) page, but I want to move some of it out of the docs.

Speaking of third party tools in general, with *.yw7* as an intermediate file and [yw2nw](https://peter88213.github.io/yw2nw/), you can use [a vast ecosystem](https://peter88213.github.io/) for importing a work in progress (full text) or an outline from various sources. 

![yw7_ecosystem](https://github.com/vkbo/novelWriter/assets/42072564/c7411aab-04c4-48d6-8709-13458e9562b4)

**Reply by vkbo** [reply]
> Also, I can't promise that I will continue to maintain the yw2nw code if the file format of novelWriter changes ever again.

The major change of the file format was for release 2.0, and it has now been generalised in a way that should prevent needs for major restructuring in the future. In particular the settings section which now supports the three main data structures (value, dictionary and class serialisation) in a standard way. I will probably add fields whenever there are new features that require it, but pretty much everything has default values.

novelWriter is currently able to read all previous formats that have been released, and I have no intention of changing that. The test suite tests against all formats to date as well to make sure they keep working.

The only potential need to update format support is to pass on new values for new, compatible features.

**Reply by vkbo** [reply]
In one of the next releases, I will add a single file format for novelWriter projects though, but it will merely be a zip file of the project folder, so supporting it should be trivial. It will also be optional, and the project folder format will still be supported.

I haven't decided if I want to give it a new file extension, or just stick with `.nwx` and just detect the file type when opening. Probably the latter.

**Reply by peter88213** [reply]
Alright, now I did program a [Python script](https://github.com/peter88213/mm2nw) that creates a novelWriter project without an intermediate *.yw7* file.

**Reply by vkbo** [reply]
Very nice! 😃
